controllers/consoleplugin_reconciler.go

```diff
@@ -0,0 +1,128 @@
+package controllers
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-logr/logr"
+	consolev1 "github.com/openshift/api/console/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/env"
+	"k8s.io/utils/ptr"
+	ctrlruntime "sigs.k8s.io/controller-runtime"
+
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+
+	"github.com/kuadrant/kuadrant-operator/pkg/library/reconcilers"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+	"github.com/kuadrant/kuadrant-operator/pkg/log"
+	"github.com/kuadrant/kuadrant-operator/pkg/openshift"
+	"github.com/kuadrant/kuadrant-operator/pkg/openshift/consoleplugin"
+)
+
+//+kubebuilder:rbac:groups=console.openshift.io,resources=consoleplugins,verbs=get;list;watch;create;update;patch;delete
+
+var (
+	ConsolePluginImageURL = env.GetString("RELATED_IMAGE_CONSOLEPLUGIN", "quay.io/kuadrant/console-plugin:latest")
+)
+
+type ConsolePluginReconciler struct {
+	*reconcilers.BaseReconciler
+
+	namespace string
+}
+
+func NewConsolePluginReconciler(mgr ctrlruntime.Manager, namespace string) *ConsolePluginReconciler {
+	return &ConsolePluginReconciler{
+		BaseReconciler: reconcilers.NewBaseReconciler(
+			mgr.GetClient(), mgr.GetScheme(), mgr.GetAPIReader(),
+			log.Log.WithName("consoleplugin"),
+			mgr.GetEventRecorderFor("ConsolePlugin"),
+		),
+		namespace: namespace,
+	}
+}
+
+func (r *ConsolePluginReconciler) Subscription() *controller.Subscription {
+	return &controller.Subscription{
+		ReconcileFunc: r.Run,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: ptr.To(openshift.ConsolePluginGVK.GroupKind())},
+			{
+				Kind:            ptr.To(ConfigMapGroupKind),
+				ObjectNamespace: r.namespace,
+				ObjectName:      TopologyConfigMapName,
+				EventType:       ptr.To(controller.CreateEvent),
+			},
+			{
+				Kind:            ptr.To(ConfigMapGroupKind),
+				ObjectNamespace: r.namespace,
+				ObjectName:      TopologyConfigMapName,
+				EventType:       ptr.To(controller.DeleteEvent),
+			},
+		},
+	}
+}
+
+func (r *ConsolePluginReconciler) Run(eventCtx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, _ *sync.Map) error {
+	logger := r.Logger()
+	logger.V(1).Info("task started")
+	ctx := logr.NewContext(eventCtx, logger)
+
+	existingTopologyConfigMaps := topology.Objects().Items(func(object machinery.Object) bool {
+		return object.GetName() == TopologyConfigMapName && object.GetNamespace() == r.namespace && object.GroupVersionKind().Kind == ConfigMapGroupKind.Kind
+	})
+
+	topologyExists := len(existingTopologyConfigMaps) > 0
+
+	// Service
+	service := consoleplugin.Service(r.namespace)
+	if !topologyExists {
+		utils.TagObjectToDelete(service)
+	}
+	err := r.ReconcileResource(ctx, &corev1.Service{}, service, reconcilers.CreateOnlyMutator)
```

Comment:
I guess if we want to align with the conventions agreed, we do not want to use this `r.ReconcileResource` function, since it performs a `Get` from the api server. Instead we want to get this resource from the topology 🤔 
Although not sure do we want to block this PR for this

Reply:
`ReconcileResource` has some functionality we still does not have in the new approach. Do you mind letting this implementation go through? In exchange, I will open issue to address that as tech debt

BTW, I realize that the old issue of endless reconciliation loop for authconfig https://github.com/Kuadrant/kuadrant-operator/issues/354 we fixed with the dry run https://github.com/Kuadrant/kuadrant-operator/blob/main/pkg/library/reconcilers/base_reconciler.go#L156

```go
desired.SetResourceVersion(obj.GetResourceVersion())
	if err := b.Client().Update(ctx, desired, client.DryRunAll); err != nil {
		return err
	}
```
can now become an issue again with the topology being the source of truth.

Reply:
added

Reply:
Yeah, I'm fine with leaving it to be addressed later 👍

Reply:
Created issue https://github.com/Kuadrant/kuadrant-operator/issues/916